### PR TITLE
Add new `Lint/NonAtomicFileOperation` cop

### PR DESCRIPTION
**Pull request by ydah**
This PR is add new `Lint/NonAtomicFileOperation` cop

Checks for non-atomic file operations.
And then replace it with a nearly equivalent and atomic method.

These can cause problems that are difficult to reproduce, especially in cases of frequent file operations in parallel, such as test runs with parallel_rspec.

For examples: creating a directory if there is none, has the following problems

An exception occurs when the directory didn't exist at the time of `exist?`, but someone else created it before `mkdir` was executed.

Subsequent processes are executed without the directory that should be there when the directory existed at the time of `exist?`, but someone else deleted it shortly afterwards.

## @safety
This cop is unsafe, because autocorrection change to atomic processing.
The atomic processing of the replacement destination is not guaranteed to be strictly equivalent to that before the replacement.

## @example
```ruby
# bad
unless FileTest.exist?(path)
  FileUtils.makedirs(path)
end

if FileTest.exist?(path)
  FileUtils.remove(path)
end

# good
FileUtils.mkdir_p(path)

FileUtils.rm_rf(path)
```
-----------------

Before submitting the PR make sure the following are checked:

* [x] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [x] Wrote [good commit messages][1].
* [-] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [x] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.

[1]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

**Comment by bbatsov** [comment]
In this case the `Redundant` check is actually `Dangerous` if you ask me, so perhaps we can rename the cop accordingly. Something like `NonAtomicFileOperation` might also be appropriate IMO. @rubocop/rubocop-core what do you think? 

P.S. Naming is still hard! :D

**Comment by ydah** [comment]
Naming always annoys me for a long time...
I feel that `NonAtomicFileOperation` looks good and I changed it. What do you think?

**Comment by bbatsov** [comment]
It's fine by me, but I'll leave this open for a couple of days to see if some of the other maintainers comes up with a better name. 

It'd be nice if we also mention this in the Ruby style guide, as I doubt many people are aware of this problem.

File: config/default.yml [hunk 1]
```diff
@@ -1956,6 +1956,13 @@ Lint/NoReturnInBeginEndBlocks:
   Enabled: pending
   VersionAdded: '1.2'
 
+Lint/NonAtomicFileOperation:
+  Description: Checks for non-atomic file operations.
+  Enabled: pending
+  VersionAdded: '<<next>>'
+  Safe: false
```
Comment (koic) [inline config/default.yml:1963]: I think this `Safe: false` can be removed because it seems likely that false positives will not occur.

```suggestion
```

**Comment by mathieujobin** [comment]
this new cop makes my rubocop crash...
on both 1.30.0 and 1.30.1

image

**Comment by koic** [comment]
@mathieujobin Can you open a new issue with the bug report template?
https://github.com/rubocop/rubocop/issues/new/choose

**Pull request merged**